collector/src/test/java/com/bitmovin/analytics/retryBackend/RetryBackendTest.kt

```diff
@@ -0,0 +1,188 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.net.Uri
+import android.os.Handler
+import com.bitmovin.analytics.BitmovinAnalyticsConfig
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.DeviceInformation
+import com.bitmovin.analytics.data.EventData
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.times
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import java.net.SocketTimeoutException
+import java.util.Calendar
+import java.util.Date
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.thread
+import okhttp3.Call
+import okhttp3.Callback
+import org.assertj.core.api.Assertions
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers.anyLong
+import org.mockito.Mockito
+import org.powermock.core.classloader.annotations.PowerMockIgnore
+import org.powermock.core.classloader.annotations.PrepareForTest
+import org.powermock.modules.junit4.PowerMockRunner
+import org.powermock.reflect.Whitebox
+
+@RunWith(PowerMockRunner::class)
+@PrepareForTest(Uri::class, RetryBackend::class, Handler::class)
+@PowerMockIgnore("javax.net.ssl.*")
+class RetryBackendTest {
+
+    private val config = BitmovinAnalyticsConfig()
+    private val backendMock = mock<Backend>()
+    private val callbackMock = mock<Callback>()
+    private val callMock = mock<Call>()
+    private val handlerMock = mock<Handler>()
+    private val deviceInformation = DeviceInformation("manufacturer", "model", false, "userAgentString", "locale", "packageName", 0, 0)
+
+    private val firstDate = Date()
+
+    private val secondDate = Calendar.getInstance().run {
+        add(Calendar.HOUR, 2)
+        time
+    }
+
+    private val thirdDate = Calendar.getInstance().run {
+        add(Calendar.HOUR, 3)
+        time
+    }
+
+    @Test
+    fun sampleShouldBeProcessedAfterHttpRequestTimeout() {
+
+        whenever(backendMock.send(any(), any())).thenAnswer {
+            (it.arguments[1] as Callback).onFailure(callMock, SocketTimeoutException("Timeout"))
+        }
+
+        val retryBacked = Mockito.spy(RetryBackend(backendMock, handlerMock))
+        retryBacked.send(setupEventData(1), callbackMock)
+
+        verify(retryBacked, times(1)).processQueuedSamples()
+    }
+
+    @Test
+    fun sampleShouldBeOrderedByScheduledTime() {
+
+        val retryBacked = Mockito.spy(RetryBackend(backendMock, handlerMock))
+
+        val firstSample = setupEventData(1)
+        val secondSample = setupEventData(2)
+
+        retryBacked.addSample(RetrySample(firstSample, null, 0, thirdDate, 0))
+        retryBacked.addSample(RetrySample(secondSample, null, 0, firstDate, 0))
+
+        // wait to be sure that processing of samples is finished
+        TimeUnit.SECONDS.sleep(2)
+        var sample = Whitebox.invokeMethod<RetrySample>(retryBacked, "getSample")
```

Comment:
Don't you think this might be an indication that the class can be broken down further to avoid having to mock internals here?
What if the queue were to be moved outside of the `RetryBackend`? Wouldn't that simplify also testing and the synchronization logic?